File: java/libgsc/src/core/Gn2htrans.java
package core;

import stmp.Stmp.Ret;
import actor.N2H;

import com.google.protobuf.GeneratedMessage.Builder;
import com.google.protobuf.Message;
import com.google.protobuf.ProtocolMessageEnum;

/**
 * 
 * N2H网络事务抽象.
 *
 * @Created on: 2015年1月8日 下午7:06:11
 * @Author: xuzewen
 * 
 */
public class Gn2htrans
{
	/** 连接延迟关闭(毫秒). */
	public int lzmsec;
	/** 事务发起者. */
	public N2H n2h;
	/** 消息命令字. */
	public ProtocolMessageEnum cmd;
	/** 事务id. */
	public int tid;
	/** 事务上BEGIN消息. */
	public Message begin;
	/** BEGIN消息上的扩展数据. */
	public byte[] extbegin;
	/** END. */
	public Gend end = new Gend();

	public Gn2htrans(N2H n2h, ProtocolMessageEnum cmd, int tid, Message begin, byte[] ext)
	{
		this.n2h = n2h;
		this.cmd = cmd;
		this.tid = tid;
		this.begin = begin;
		this.extbegin = ext;
		this.lzmsec = 0;
	}

	private final void end(short ret, Builder<?> end, byte[] extend)
	{
		this.end.ret = ret;
		this.end.end = end == null ? null : end.build();
		this.end.extend = extend;
		if (this.end.ret == (short) Ret.RET_FAILURE.ordinal()) /* 失败, 关闭连接. */
		{
			this.n2h.future(v -> {
				this.n2h.close();
				this.finish();
			});
			return;
		}
		this.n2h.future(v -> {
			if (this.lzmsec > 0)
			{
				this.n2h.lts += this.lzmsec;
				this.n2h.lazyClose(); /* 延迟关闭. */
			}
			if (this.n2h.gusr != null) /* 有应用层关联. */
			{
				this.n2h.gusr.future(vx -> {
					this.n2h.gusr.sendEnd0(this);
				});
			} else
			{
				this.n2h.sendEnd(this);
			}
		});
	}

	public final void end(ProtocolMessageEnum ret, Builder<?> end, byte[] extend)
	{
		this.end((short) ret.getNumber(), end, extend);
	}

	public final void end(short ret)
	{
		this.end(ret, null, null);
	}

	public final void end(ProtocolMessageEnum ret)
	{
		this.end(ret, null, null);
	}

	public final void end(Builder<?> end)
	{
		this.end(Ret.RET_SUCCESS, end, null);
	}

	public final void end(ProtocolMessageEnum ret, Builder<?> end)
	{
		this.end(ret, end, null);
	}

	public final void success()
	{
		this.end(Ret.RET_SUCCESS, null, null);
	}

	public final void failure()
	{
		this.end(Ret.RET_FAILURE, null, null);
	}

	public final void successLazyClose(int sec)
	{
		this.lzmsec = (int) (sec == 0 ? Cfg.libgsc_n2h_lazy_close/* 默认值. */: sec * 1000);
		this.end(Ret.RET_SUCCESS, null, null);
	}

	public final void successLazyClose(Builder<?> end, int sec)
	{
		this.lzmsec = (int) (sec == 0 ? Cfg.libgsc_n2h_lazy_close/* 默认值. */: sec * 1000);
		this.end(Ret.RET_SUCCESS, end, null);
	}

	public final void successLazyClose(ProtocolMessageEnum ret, int sec)
	{
		this.lzmsec = (int) (sec == 0 ? Cfg.libgsc_n2h_lazy_close/* 默认值. */: sec * 1000);
		this.end(ret, null, null);
	}

	public final void finish()
	{
		if (Gsc.ec != null)
			Gsc.ec.gn2hTrans(this);
	}
}

File: c/libgsc/core/Gn2htrans.cpp
/*
 * Gn2htrans.cpp
 *
 *  Created on: Jan 31, 2015
 *      Author: root
 */

#include "Gn2htrans.h"
#include "Gsc.h"
#include "Cfg.h"

Gn2htrans::Gn2htrans(N2H* n2h, ushort cmd, uint tid, Message* begin, string* extbegin)
{
	this->lzmsec = 0;
	this->cmd = cmd;
	this->tid = tid;
	this->begin = begin;
	this->extbegin = (extbegin == NULL ? NULL : new string(*extbegin));
	this->ret = RET_SUCCESS;
	this->endx = NULL;
	this->extend = NULL;
	n2h->ref();
	this->n2h = n2h;
}

void Gn2htrans::end(ushort ret, Message* end, string* extend)
{
	this->ret = ret;
	this->endx = end;
	this->extend = extend == NULL ? NULL : new string(*extend);
	if (ret == RET_FAILURE) /** 失败, 关闭连接. */
	{
		this->n2h->future([this]
		{
			this->n2h->close();
			this->finish();
		});
		return;
	}
	this->n2h->future([this]
	{
		if(this->lzmsec > 0)
		{
			this->n2h->lts += this->lzmsec;
			this->n2h->lazyClose(); /** 延迟关闭. */
		}
		if (this->n2h->gusr != NULL) /** 有应用层关联. */
		{
			this->n2h->gusr->future([this]
					{
						this->n2h->gusr->sendEnd0(this);
					});
		} else
		{
			this->n2h->sendEnd(this);
		}
	});
}

void Gn2htrans::end(Message* end, string* extend)
{
	this->end(RET_SUCCESS, end, extend);
}

void Gn2htrans::end(ushort ret, string* extend)
{
	this->end(ret, NULL, extend);
}

void Gn2htrans::success(string* extend)
{
	this->end(RET_SUCCESS, NULL, extend);
}

void Gn2htrans::failure()
{
	this->end(RET_FAILURE, NULL, NULL);
}

void Gn2htrans::finish()
{
	if (Gsc::ec)
		Gsc::ec->gn2hTrans(this);
	delete this;
}

void Gn2htrans::successLazyClose(int sec)
{
	this->lzmsec = (sec == 0 ? Cfg::libgsc_n2h_lazy_close /** 默认值. */: sec * 1000);
	this->end(RET_SUCCESS, NULL, extend);
}

void Gn2htrans::endLazyClose(Message* end, int sec)
{
	this->lzmsec = (sec == 0 ? Cfg::libgsc_n2h_lazy_close /** 默认值. */: sec * 1000);
	this->end(RET_SUCCESS, end, NULL);
}

void Gn2htrans::endLazyClose(ushort ret, int sec)
{
	this->lzmsec = (sec == 0 ? Cfg::libgsc_n2h_lazy_close /** 默认值. */: sec * 1000);
	this->end(ret, NULL, NULL);
}

Gn2htrans::~Gn2htrans()
{
	this->n2h->unRef(); /** N2H在事务构造时被引用1次, 在事务结束时释放引用. */
	if (this->begin != NULL)
		delete this->begin;
	if (this->extbegin != NULL)
		delete this->extbegin;
	if (this->endx != NULL)
		delete this->endx;
	if (this->extend != NULL)
		delete this->extend;
}

